### ISSUE TITLE
Activation functions - learnable params, PReLU, Maxout etc

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#2539]
Recent PRs refactored activation functions, which was the first step for supporting activation functions with learnable parameters. The interface exists: now just has to be hooked into the parameter initializers, updaters, gradient/param views, etc.

Remaining tasks:
- Properly support activation functions with learnable parameters in DL4J
- Implement the following: (first 4 are in Keras)
    - PReLU
    - SReLU
    - Parametric softplus
    - ThresholdedReLU (no params, but is in Keras hence needed for model import)
    - Maxout

**Comment by maxpumperla** [comment]
Hey @AlexDBlack, since this is pretty old and I've been assigned for some reason, could you please give an update on where we are regarding the above list? thanks!

**Comment by raver119** [comment]
Point me to whatever you'll need on native side for this.

**Comment by agibsonccc** [comment]
Ah, a lot of this was related to keras I believe. In general, this would be low priority for you but one of the things we will need to do with keras is also reach feature parity on some fronts. You can prioritize this lower than getting some of the basics out of the way. You're also welcome to split the work on this too.

**Comment by turambar** [comment]
@agibsonccc @maxpumperla yeah this is really about feature parity with keras and some of the other frameworks. It's a nice-to-have-but-not-urgent.

**Comment by AlexDBlack** [comment]
Yup. Will require work on the DL4J side, and a bit on the ND4J side (los functions and actual implementation of  these). Not too hard in principle, the API is already there. In DL4J, just need to build out support for them in:
(a) parameter initializers
(b) layers (backprop, serialization, etc)

I agree it's not a high priority. Let me know when you want to build this and either I can add it, or I can guide you (or someone else) on how to do it.

**Comment by maxpumperla** [comment]
thanks @raver119, will do once it becomes more active/relevant

**Comment by maxpumperla** [comment]
awesome @AlexDBlack, I'll get back to you once I have a clearer picture on how to best split down and prioritise this and other related tickets.

**Comment by maxpumperla** [comment]
@AlexDBlack 

Note: 
- SReLU is gone in keras 2
- Parametric softplus is gone in keras 2.
- MaxoutDense has been deprecated in keras 2 (alongside highway networks and masking)

Might be useful for prioritisation, I think PReLU would be cool, ThreshReLU wouldn't be too hard and the other three would only be for legacy purposes if we only speak about keras compatibility.

**Comment by wanjizheng** [comment]
Any updates on this issue?

**Comment by AlexDBlack** [comment]
@wanjizheng No updates. It's not a priority at the moment. We may revisit in the future, but there hasn't been much demand from the community for these.

**Comment by maxpumperla** [comment]
just going through some old keras issues. @raver119 does your gang have capacity to implement PReLU, SReLU, parametric softplus or ThresholdedReLU in libnd4j?

**Comment by raver119** [comment]
Yes, we will start implementing those after 1.0.0-beta goes out with all perf fixes.
